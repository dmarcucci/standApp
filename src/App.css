.App {
  text-align: center;
  font-size: 14px;
  font-family: Arial, Helvetica, sans-serif;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 40vmin;
}

.App-header {
  background-color: #f55e61;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: #f6cd4c;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
