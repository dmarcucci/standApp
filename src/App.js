import React, { Component } from 'react';
import './App.css';
import { Form, FormGroup, Label, Input, Row, Col, Button, Alert } from "reactstrap";
import Calendar from "react-calendar";
import cookie from 'react-cookies';


class App extends Component {

    constructor (props) {
        super(props);

        this.state = {
            apiKey: cookie.load('apiKey'),
            token: cookie.load('token'),
            boardId: cookie.load('boardId'),
            username: cookie.load('username'),
            idMember: '',
            fullName: '',
            date: new Date(),
            data: [],
            errors: []
        };

        this.handleChange = this.handleChange.bind(this);
        this.generateStandUp = this.generateStandUp.bind(this);
        this.filterResults = this.filterResults.bind(this);
    }


    handleChange(event) {
        this.setState({[event.target.name]: event.target.value});
        cookie.save(event.target.name, event.target.value, { path: '/' })
    }

    handleChangeDate = date => this.setState({ date })

    generateStandUp() {
        let errorsTmp = []

        if (!this.state.apiKey) {
            errorsTmp.push('ApiKey is required')
        }
        if (!this.state.token) {
            errorsTmp.push('Token is required')
        }
        if (!this.state.boardId) {
            errorsTmp.push('BoardId is required')
        }
        if (!this.state.username) {
            errorsTmp.push('Username is required')
        }

        this.setState({
            errors: errorsTmp
        })


        if (errorsTmp.length === 0) {
            fetch('https://api.trello.com/1/members/' + this.state.username)
                .then(response => response.json())
                .then(
                    (data) => {
                        this.setState({
                            idMember: data.id,
                            fullName: data.fullName
                        }, () => {
                            fetch('https://api.trello.com/1/boards/' + this.state.boardId +
                                '/actions?key=' + this.state.apiKey + '&token=' + this.state.token)
                                .then(response => {
                                    if (response.status !== 200) {
                                        throw new Error(response.status)
                                    }
                                    return response.json()
                                })
                                .then(data => {
                                    let filteredData = this.filterResults(data)
                                    this.setState({data: filteredData})
                                }, (error) => {
                                    let errorMsg
                                    let errorCode = error.toString().match(/\d+/).map(Number)[0]

                                    switch (errorCode) {
                                        case 400:
                                            errorMsg = 'Invalid Board Id'
                                            break
                                        case 401:
                                            errorMsg = 'Wrong credentials. Access denied'
                                            break
                                        default:
                                            errorMsg = 'Unknown error'
                                    }

                                    this.setState({
                                        errors: [errorMsg]
                                    })
                                });
                        })
                    },
                    (error) => {
                        this.setState({
                            errors: ['Username does not exist']
                        })
                    }
                );
        }
    }

    filterResults(data) {
        let filteredData = data.filter((item) => {
            let convDate = new Date(new Date(item.date).toLocaleString("en-US"));
                return item.idMemberCreator === this.state.idMember &&
                convDate.getDate() === this.state.date.getDate() &&
                convDate.getMonth() === this.state.date.getMonth() &&
                convDate.getFullYear() === this.state.date.getFullYear()
        });

        return filteredData
    }

    render() {
        const errorMsgs = this.state.errors.map((msg, id) => {
            return (<Alert key={id} color="danger">{msg}</Alert>)
        })

        return (
            <div className="App">
                <header className="App-header">
                    StandApp
                </header>
                <br/><br/>
                <Row className="justify-content-center">
                <Col sm="4">
                    <div>
                        {errorMsgs}
                    </div>
                    <Form>
                        <FormGroup>
                            <Label for="apiKey">API Key</Label>
                            <Input type="textfield" name="apiKey" id="apiKey" placeholder="Enter your API Key" value={this.state.apiKey} onChange={this.handleChange}/>
                        </FormGroup>
                        <FormGroup>
                            <Label for="token">Token</Label>
                            <Input type="password" name="token" id="token" placeholder="Enter your Token" value={this.state.token} onChange={this.handleChange}/>
                        </FormGroup>
                        <FormGroup>
                            <Label for="token">Board ID</Label>
                            <Input type="textfield" name="boardId" id="boardId" placeholder="Enter Board ID" value={this.state.boardId} onChange={this.handleChange}/>
                        </FormGroup>
                        <FormGroup>
                            <Label for="username">Username</Label>
                            <Input type="textfield" name="username" id="username" placeholder="Enter Username" value={this.state.username} onChange={this.handleChange}/>
                        </FormGroup>
                        <FormGroup>
                            <Label for="calendar">StandUp Day</Label>
                            <Calendar
                                name="calendar"
                                onChange={this.handleChangeDate}
                                value={this.state.date}
                            />
                        </FormGroup>
                        <Button onClick={() => this.generateStandUp()}>Generate Stand Up</Button>
                    </Form>
                </Col>
                </Row>
                <br/><br/>
                <Row className="justify-content-center">
                    {(this.state.data.length > 0) ? <div><div>Keep it up {this.state.fullName}!</div>
                    <div><pre>{JSON.stringify(this.state.data, null, 2) }</pre></div></div> : (this.state.fullName) ? <div>You have been lazy {this.state.fullName}!</div> : null}
                </Row>
            </div>
        );
    }
}

export default App;
