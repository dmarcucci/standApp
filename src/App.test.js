import React from 'react';
import { shallow, mount } from 'enzyme';
import App from './App';
import { FormGroup, Label, Input, Button, Alert} from 'reactstrap';
import Calendar from "react-calendar";

describe('App', () => {
    var MockDate = require('mockdate');

    it('should render component correctly', () => {
        MockDate.set(1434319925275);

        const component = shallow(<App />);
        expect(component).toMatchSnapshot();

        MockDate.reset();
    });

    it('form submission with missing parameters', () => {
        const wrapper = mount(<App />);

        const button = wrapper.find(Button);
        const errorMsg = wrapper.find('.errorMsg');

        expect(button.length).toBe(1);
        expect(errorMsg.length).toBe(1);
        expect(wrapper.state('errors').length).toBe(0);
        expect(wrapper.find(Alert).length).toBe(0);

        button.simulate('click');
        expect(wrapper.state('errors').length).toBe(3);
        expect(wrapper.find(Alert).length).toBe(3);

        wrapper.state().errors.forEach((error, id) => {
            expect(wrapper.contains(error)).toBe(true);
        });
    });

    //not working yet
    it('form submission with wrong credentials', () => {
        const wrapper = mount(<App />);

        const inputApiKey = wrapper.find('input#apiKey').getDOMNode();
        const inputToken = wrapper.find('input#token').getDOMNode();
        const devTeam = wrapper.find('input#devTeam').getDOMNode();
        const button = wrapper.find(Button);
        const errorMsg = wrapper.find('.errorMsg');

        expect(wrapper.state('errors').length).toBe(0);
        expect(wrapper.find(Alert).length).toBe(0);

        inputApiKey.value = "wrong key";
        inputToken.value = "wrong token";
        devTeam.value = "Chameleon";

        button.simulate('click');

        expect(wrapper.state('errors').length).toBe(2);
        expect(wrapper.find(Alert).length).toBe(2);

        wrapper.state().errors.forEach((error, id) => {
            expect(wrapper.contains(error)).toBe(true);
        });
    });
});

