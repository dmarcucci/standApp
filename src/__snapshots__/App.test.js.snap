// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render component correctly 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    StandApp
  </header>
  <br />
  <br />
  <Row
    className="justify-content-center"
    tag="div"
  >
    <Col
      sm="4"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <div
        className="errorMsg"
      />
      <Form
        tag="form"
      >
        <FormGroup
          tag="div"
        >
          <Label
            for="apiKey"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            API Key
          </Label>
          <Input
            id="apiKey"
            name="apiKey"
            onChange={[Function]}
            placeholder="Enter your API Key"
            type="textfield"
          />
        </FormGroup>
        <FormGroup
          tag="div"
        >
          <Label
            for="token"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Token
          </Label>
          <Input
            id="token"
            name="token"
            onChange={[Function]}
            placeholder="Enter your Token"
            type="password"
          />
        </FormGroup>
        <Label
          for="devTeam"
          tag="label"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          Team
        </Label>
        <Autocomplete
          autoHighlight={true}
          getItemValue={[Function]}
          inputProps={
            Object {
              "class": "form-control",
              "id": "devTeam",
              "name": "devTeam",
              "placeholder": "Enter Your Team",
            }
          }
          isItemSelectable={[Function]}
          items={
            Array [
              Object {
                "label": "Chameleon",
              },
              Object {
                "label": "Codebusters",
              },
              Object {
                "label": "Scrumdiddy",
              },
            ]
          }
          menuStyle={
            Object {
              "background": "rgba(255, 255, 255, 0.9)",
              "borderRadius": "3px",
              "boxShadow": "0 2px 12px rgba(0, 0, 0, 0.1)",
              "fontSize": "90%",
              "maxHeight": "50%",
              "overflow": "auto",
              "padding": "2px 0",
              "position": "fixed",
            }
          }
          onChange={[Function]}
          onMenuVisibilityChange={[Function]}
          onSelect={[Function]}
          renderInput={[Function]}
          renderItem={[Function]}
          renderMenu={[Function]}
          selectOnBlur={false}
          shouldItemRender={[Function]}
          value=""
          wrapperProps={
            Object {
              "class": "form-group",
            }
          }
          wrapperStyle={
            Object {
              "display": "block",
            }
          }
        />
        <FormGroup
          tag="div"
        >
          <Label
            for="token"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Board
          </Label>
          <Input
            disabled="disabled"
            id="boardId"
            name="boardId"
            onChange={[Function]}
            type="select"
          >
            <option
              value=""
            >
              Select a Board
            </option>
          </Input>
        </FormGroup>
        <FormGroup
          tag="div"
        >
          <Label
            for="calendar"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            StandUp Day
          </Label>
          <Calendar
            maxDetail="month"
            minDetail="century"
            name="calendar"
            onChange={[Function]}
            returnValue="start"
            showNavigation={true}
            showNeighboringMonth={true}
            value={2015-06-14T22:12:05.275Z}
            view="month"
          />
        </FormGroup>
        <Button
          color="secondary"
          onClick={[Function]}
          tag="button"
        >
          Generate Stand Up
        </Button>
      </Form>
    </Col>
  </Row>
  <br />
  <br />
  <Row
    className="justify-content-center"
    tag="div"
  />
</div>
`;
